util/base.hpp

```diff
@@ -0,0 +1,350 @@
+
+    end.t = goalTime;
+    end.pos = p;
+
+    createFromWaypoints({start, end}, isBaseFrame);
+  }
+
+  static CartesianGoal createFromWaypoints(std::vector<Waypoint> waypoints, bool isBaseFrame) {
+    auto count = waypoints.size();
+
+    VectorXd times(count);
+    MatrixXd positions(3, count);
+    MatrixXd velocities(3, count);
+    MatrixXd accelerations(3, count);
+
+    for (auto idx = 0; idx < count; ++idx) {
+      times(idx) = waypoints.at(idx).t;
+      positions(0, idx) = waypoints.at(idx).pos.x;
+      positions(1, idx) = waypoints.at(idx).pos.y;
+      positions(2, idx) = waypoints.at(idx).pos.theta;
+
+      velocities(0, idx) = waypoints.at(idx).vel.x;
+      velocities(1, idx) = waypoints.at(idx).vel.y;
+      velocities(2, idx) = waypoints.at(idx).vel.theta;
+
+      accelerations(0, idx) = waypoints.at(idx).acc.x;
+      accelerations(1, idx) = waypoints.at(idx).acc.y;
+      accelerations(2, idx) = waypoints.at(idx).acc.theta;
+    }
+
+    return CartesianGoal(times, positions, velocities, accelerations);
+  }
+
+  // Creates from cartesian trajectory.  May be (x,y) or (x,y,theta), but 
+  // (x,y,theta) less likely to be realizable depending on system.
+  // If "theta" is not defined, is tangent to path.
+  // (x, y path here could be used for planning path through obstacles, for example)
+  // \param isBaseFrame is "true" if this is relative to current body frame,
+  // and "false" for using odometry coordinates.
+  // TODO
+  static CartesianGoal createFromTrajectory(CartesianTrajectory t, bool isBaseFrame);
+
+  const Eigen::VectorXd& times() const { return times_; }
+  const Eigen::Matrix<double, 3, Eigen::Dynamic>& positions() const { return positions_; }
+  const Eigen::Matrix<double, 3, Eigen::Dynamic>& velocities() const { return velocities_; }
+  const Eigen::Matrix<double, 3, Eigen::Dynamic>& accelerations() const { return accelerations_; }
+
+private:
+  CartesianGoal(const Eigen::VectorXd& times,
+       const Eigen::Matrix<double, 3, Eigen::Dynamic>& positions,
+       const Eigen::Matrix<double, 3, Eigen::Dynamic>& velocities,
+       const Eigen::Matrix<double, 3, Eigen::Dynamic>& accelerations)
+    : times_(times),
+      positions_(positions),
+      velocities_(velocities),
+      accelerations_(accelerations) {}
+
+  // TODO: some info here capturing above states/options
+  const Eigen::VectorXd times_{0};
+  const Eigen::Matrix<double, 3, Eigen::Dynamic> positions_{0, 0};
+  const Eigen::Matrix<double, 3, Eigen::Dynamic> velocities_{0, 0};
+  const Eigen::Matrix<double, 3, Eigen::Dynamic> accelerations_{0, 0};
+
+  bool local_trajectory_{};
+};
+
+//////////////////////////////
+
+class MobileBase {
+
+public:
+
+  // Parameters for creating a base
+  struct Params: public GroupManager::Params {
+  };
+
+  MobileBase(Params p) :
+    group_manager_(GroupManager::create(p))
+  {}
+
+  // This is a generic implementation that assumes the use of a wheel-space trajectory
+  bool update() {
```

Comment:
This is the function that makes me question the current separation/use of the `GroupManager`.  Most of the logic in it is for dealing with joint-space trajectories; however, here we are not really keeping the trajectory in the group manager at all.  I feel like we're kind of misusing it there as a simple group wrapper.

Reply:
See current attempt to move this over to "GroupTrajectoryManager" with GroupManager being a simpler class without the trajectory state